### PR TITLE
Implemented calendar functionality

### PR DESCRIPTION
**Pull request by kristen-m**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,85 @@
...
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  private static int DAY_LENGTH = 24*60;
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    Set<String> attendees =  new HashSet<String>();
+    int meetingDur = (int)request.getDuration();
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit. add a space
`(int) request`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,85 @@
...
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  private static int DAY_LENGTH = 24*60;
```
Comment (Kaleab-Gebremichael) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: There is a WHOLE_DAY const in the TimeRange so you can instead compare the request.getDuration() to TimeRange.WHOLE_DAY.getDuration()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,85 @@
...
+    int meetingDur = (int)request.getDuration();
+    attendees.addAll(request.getAttendees());
+    if (meetingDur > DAY_LENGTH){
+      return possibleTimes;
+    }
+    if (events.isEmpty() || attendees.isEmpty()){
+      possibleTimes.add(TimeRange.WHOLE_DAY);
+      return possibleTimes;
+    }
+    /*A time is considered to be busy when an attendee is in an event & the meeting request*/
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: For inline comments, use `\\` instead of javaDoc style `/* ... */`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,85 @@
...
+      if (blockedTimes[i].overlaps(busyWithOverlap.get(index))){
+        int meetingStart = Math.min(busyWithOverlap.get(index).start(), blockedTimes[i].start());
+        int meetingEnd = Math.max(busyWithOverlap.get(index).end(), blockedTimes[i].end()) - 1;
+        busyWithOverlap.set(index, TimeRange.fromStartEnd(meetingStart, meetingEnd, true));
+      }
+      else {
+        busyWithOverlap.add(blockedTimes[i]);
+        index ++;
+      }
+    }
```
Comment (Kaleab-Gebremichael) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It looks like this could be a good chunk for a helper method!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,85 @@
...
+    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    for(Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      Set<String> overlap = new HashSet<>(eventAttendees);
+      overlap.retainAll(attendees);
+      if (!overlap.isEmpty()){
+        busyTimes.add(event.getWhen());
+      }
+    }
+  /*Combines overlapping busy times into an ArrayList*/
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: same here, also in general -- these types of comments which describe an upcoming block of code, are often a good sign that this could benefit from a helper function.

Then ideally, instead of needing to describe what is happening, the method name can be descriptive enough to aid in understanding

**Review by Kaleab-Gebremichael — APPROVED** [review]
Looks really good!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,85 @@
...
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    TimeRange[] blockedTimes = ((List<TimeRange>) busyTimes).toArray(new TimeRange[busyTimes.size()]);
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Why is this needed? Can you just use busyTimes as is?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,85 @@
...
+    int index = 0;
+    for (int i=0; i < blockedTimes.length; i++){
+      if (blockedTimes[i].overlaps(busyWithOverlap.get(index))){
+        int meetingStart = Math.min(busyWithOverlap.get(index).start(), blockedTimes[i].start());
+        int meetingEnd = Math.max(busyWithOverlap.get(index).end(), blockedTimes[i].end()) - 1;
+        busyWithOverlap.set(index, TimeRange.fromStartEnd(meetingStart, meetingEnd, true));
+      }
+      else {
+        busyWithOverlap.add(blockedTimes[i]);
+        index ++;
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit. remove space 
`index++`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -15,9 +15,85 @@
...
+      }
+    }
+    possibleTimes = findFreeTimes(busyWithOverlap, request.getDuration());
+    return possibleTimes;
   }
-}
+
+  /**
+  * Returns an ArrayList of possible meeting times that are available for all attendees
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit. end this with a period

**Review by damargulis — COMMENTED** [review]
Overall looks good! -- but I think you could use a couple helper methods to simplify things a bit and make it a bit more readable

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -42,7 +43,37 @@
       possibleTimes.add(TimeRange.WHOLE_DAY);
       return possibleTimes;
     }
-    
+
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    for(Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      Set<String> overlap = new HashSet<>(eventAttendees);
+      overlap.retainAll(attendees);
```
Comment (juliareichel) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: never seen this retainAll method- cool usage

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -42,7 +43,37 @@
...
+      if (!overlap.isEmpty()){
+        busyTimes.add(event.getWhen());
+      }
+    }
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    System.out.println(busyTimes);  
+    TimeRange[] unavailable = ((List<TimeRange>) busyTimes).toArray(new TimeRange[busyTimes.size()]);
```
Comment (juliareichel) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: often easier to save collection into arraylist because inherently an arraylist is a type of a collection so there's more flexibility than just using a regular array

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed that

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yep, I needed to go back and refresh my memory on ArrayLists in order to get the syntax right, fixed in the last PR

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Got it

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Removed these in latest PR (replaced w helper function)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: got it

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Good idea! Put it in the last PR, it definitely helps readability

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks for pointing this out!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Good to know the benefits here! I switched back to ArrayList in the earlier commits so it should look better now!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -15,9 +15,94 @@
...
+    possibleTimes = findFreeTimes(busyWithOverlap, request.getDuration());
+    return possibleTimes;
   }
-}
+
+/**
+  * Returns an ArrayList of meeting times that are unavailable due to schedule
+  * conflicts and overlapping meeting times within the given group of attendees. 
+  *
+  * @param  events the Collection of Event objects passed into the query function
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: minor nit.
These parameters don't actually have to be from the query. This function works to get the busy times given any collection of events and set of attendees.  You are only using it for this query, but there is no need to specify that in the comments here that those params must come from the query

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -15,9 +15,94 @@
...
+      if (!overlap.isEmpty()){
+        busyTimes.add(event.getWhen());
+      }
+    }
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    ArrayList<TimeRange> busyWithOverlap = new ArrayList<TimeRange>();
+    if(busyTimes.size() > 0) {
+      busyWithOverlap.add(busyTimes.get(0));
+    }
+    int index = 0;
```
Comment (damargulis) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: use something like busyWithOverlayIndex and busyTimesIndex, instead of index and i to make it clear which belongs to which

**Review by damargulis — APPROVED** [review]
Looks good! Great job on the refactor!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Good idea, changed in last PR

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (kristen-m) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: True, updated the javadoc to reflect this!
